src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -0,0 +1,449 @@
+                    {
+                        scope.Free();
+                    }
+                    NestedScopes.Free();
+
+                    foreach (var closure in Closures)
+                    {
+                        closure.Free();
+                    }
+                    Closures.Free();
+                    DeclaredCapturedVariables.Free();
+                    Environments.Free();
+                }
+            }
+
+            private sealed class Closure
+            {
+                public MethodSymbol Symbol { get; }
+
+                public PooledHashSet<Symbol> CapturedVariables { get; } = PooledHashSet<Symbol>.GetInstance();
+
+                public ArrayBuilder<LambdaFrame> CapturedEnvironments { get; } = ArrayBuilder<LambdaFrame>.GetInstance();
+
+                public NamedTypeSymbol ContainingType { get; set; }
+
+                public Closure(MethodSymbol symbol)
+                {
+                    Symbol = symbol;
+                }
+
+                public void Free()
+                {
+                    CapturedVariables.Free();
+                    CapturedEnvironments.Free();
+                }
+            }
+
+            /// <summary>
+            /// Optimizes local functions such that if a local function only references other local functions without closures, it itself doesn't need a closure.
+            /// </summary>
+            private void RemoveUnneededReferences()
+            {
+                var methodGraph = new MultiDictionary<MethodSymbol, MethodSymbol>();
+                var capturesThis = new HashSet<MethodSymbol>();
+                var capturesVariable = new HashSet<MethodSymbol>();
+                var visitStack = new Stack<MethodSymbol>();
+                VisitClosures(_scopeTree, (scope, closure) =>
+                {
+                    foreach (var capture in closure.CapturedVariables)
+                    {
+                        if (capture is MethodSymbol localFunc)
+                        {
+                            methodGraph.Add(localFunc, closure.Symbol);
+                        }
+                        else if (capture == _topLevelMethod.ThisParameter)
+                        {
+                            if (capturesThis.Add(closure.Symbol))
+                            {
+                                visitStack.Push(closure.Symbol);
+                            }
+                        }
+                        else if (capturesVariable.Add(closure.Symbol) && !capturesThis.Contains(closure.Symbol))
+                        {
+                            visitStack.Push(closure.Symbol);
+                        }
+                    }
+                });
+
+                while (visitStack.Count > 0)
+                {
+                    var current = visitStack.Pop();
+                    var setToAddTo = capturesVariable.Contains(current) ? capturesVariable : capturesThis;
```

Comment:
Yeah, I'm not convinced this code is correct. For now, though, I want to match semantics bug-for-bug.